src/fromager/wheels.py

```diff
@@ -129,6 +131,73 @@ def analyze_wheel_elfdeps(
     return requires, provides
 
 
+def add_extra_metadata_to_wheels(
+    ctx: context.WorkContext,
+    req: Requirement,
+    version: Version,
+    extra_environ: dict[str, str],
+    sdist_root_dir: pathlib.Path,
+    wheel_file: pathlib.Path,
+) -> None:
+    extra_data_plugin = overrides.find_override_method(
+        req.name, "add_extra_metadata_to_wheels"
+    )
+    data_to_add = {}
+    if extra_data_plugin:
+        data_to_add = overrides.invoke(
+            extra_data_plugin,
+            ctx=ctx,
+            req=req,
+            version=version,
+            extra_environ=extra_environ,
+            sdist_root_dir=sdist_root_dir,
+        )
+        if not isinstance(data_to_add, dict):
+            logger.warning(
+                f"{req.name}: unexpected return type from plugin add_extra_metadata_to_wheels. Expected dictionary. Will ignore"
+            )
+            data_to_add = {}
+
+    with tempfile.TemporaryDirectory() as dir_name:
+        cmd = ["wheel", "unpack", str(wheel_file), "-d", dir_name]
+        external_commands.run(
+            cmd,
+            cwd=dir_name,
+            network_isolation=ctx.network_isolation,
+        )
+
+        wheel_file_name_parts = wheel_file.name.split("-")
+        distribution_name = f"{wheel_file_name_parts[0]}-{wheel_file_name_parts[1]}"
+        dist_info_dir = (
+            pathlib.Path(dir_name)
+            / distribution_name
+            / f"{distribution_name}.dist-info"
+        )
+        if not dist_info_dir.exists():
+            logger.debug(
+                f"{req.name}: could not add additional metadata. {dist_info_dir} does not exist"
+            )
+            return
+
+        build_file = dist_info_dir / "fromager-build-settings"
+        build_file.write_text(
+            tomlkit.dumps(ctx.settings.get_package_settings(req.name))
+        )
+        if data_to_add:
+            build_file.write_text(tomlkit.dumps({"metadata-from-plugin": data_to_add}))
```

Comment:
I verified that calling write_text 2 times overwrites the contents of the file.

```
>>> import pathlib
>>> p = pathlib.Path('tmp.txt')
>>> p.write_text('foo')
3
>>> p.read_text()
'foo'
>>> p.write_text('bar')
3
>>> p.read_text()
'bar'
```

I think you want to combine the data structures before writing. Something like this:

```suggestion
        settings = ctx.settings.get_package_settings(req.name)
        if data_to_add:
            settings['metadata-from-plugin'] = data_to_add
        else:
            settings['metadata-from-plugin'] = {}
        build_file.write_text(
            tomlkit.dumps(settings)
        )
```

Reply:
Do we want an empty `metadata-from-plugin` key even if there is no data from the plugin?

Reply:
Normally I'd assume so because it makes it more consistent for something reading the file. Does that make weird looking TOML, though?

Reply:
Yes it looks like this:

```toml
[download_source]
url = "https://files.pythonhosted.org/packages/e7/c1/b210bf1071c96ecfcd24c2eeb4c828a2a24bf74b38af13896d02203b1eec/stevedore-${version}.tar.gz"
destination_filename = "stevedore-custom-${version}.tar.gz"

[resolver_dist]
sdist_server_url = "https://pypi.org/simple"
include_sdist = false
include_wheels = true

[metadata-from-plugin]
```

Reply:
So a little weird but not terrible.

I guess in other files the keys are left out if there are no settings, so maybe we should follow that pattern.